nyxtools/isara_robot.py

```diff
@@ -0,0 +1,313 @@
+
+    # Statuses
+
+    # interface from denso, try to maintain types
+    # int
+    # status
+
+    # bool
+    # busy_sts
+
+    # bool
+    # mount_ready_sts
+
+    # is mounting bool
+    # mounting_sts
+
+    # is dismounting bool
+    # dismounting_sts
+
+    # is drying bool
+    # 0 = "Idle"
+    # 1 = "Drying"
+    drying_sts = Cpt(EpicsSignalRO, "GripDry-Sts")
+    # new
+    # 0 = "Hold off"
+    # 1 = "Permit"
+    drying_permitted_sts = Cpt(EpicsSignalRO, "DryPmt-I")
+
+    # spindle occupied bool
+    spindle_occupied_sts = Cpt(EpicsSignalRO, "Samp:Dif-Sts")
+
+    # New statuses for ISARA
+
+    # Last Message
+    last_message = Cpt(EpicsSignalRO, "LastMsg-I")
+
+    # Fault Status
+    # 0 = "Ok"
+    # 1 = "Fault"
+    fault_sts = Cpt(EpicsSignalRO, "Flt-Sts")
+
+    # Alarm Status
+    # TODO: What does this return?
+    alarm_sts = Cpt(EpicsSignalRO, "Alarm-I")
+
+    # Current Position
+    position_sts = Cpt(EpicsSignalRO, "Pos:Name-I")
+
+    # Moving Status
+    # 0 = "Stopped"
+    # 1 = "Moving"
+    moving_sts = Cpt(EpicsSignalRO, "Seq:Run-Sts")
+
+    # Paused Status
+    # 0 = "Normal"
+    # 1 = "Paused"
+    paused_sts = Cpt(EpicsSignalRO, "Seq:Paus-Sts")
+
+    # Speed Status
+    speed_sts = Cpt(EpicsSignalRO, "Speed-I")
+
+    # Current Tool
+    # 0 = "ToolChanger"
+    # 1 = "Cryotong"
+    # 2 = "SingleGripper"
+    # 3 = "DoubleGripper"
+    # 4 = "MiniSpineGripper"
+    # 5 = "RotatingGripper"
+    # 6 = "PlateGripper"
+    # 7 = "Spare"
+    # 8 = "LaserTool"
+    current_tool = Cpt(EpicsSignal, "Tl-I")
+
+    # Power status
+    # 0 = "Off"
+    # 1 = "On"
+    power_sts = Cpt(EpicsSignalRO, "Pwr-Sts")
+
+    # Occupied statuses
+    # 0 = "Empty"
+    # 1 = "Present"
+    samp_a_occ_sts = Cpt(EpicsSignalRO, "Samp:A-Sts")
+    samp_b_occ_sts = Cpt(EpicsSignalRO, "Samp:B-Sts")
+    samp_dif_occ_sts = Cpt(EpicsSignalRO, "Samp:Dif-Sts")
+
+    # Gripper Statuses
+    # 0 = "Open"
+    # 1 = "Closed"
+    grip_a_sts = Cpt(EpicsSignalRO, "Grp:A-Sts")
+    grip_b_sts = Cpt(EpicsSignalRO, "Grp:B-Sts")
+
+    # Samples occupying gripper/spindle
+    # returns 0-29
+    # returns -1 if empty
+    puck_a_read = Cpt(EpicsSignalRO, "Pck:A-I")
+    puck_b_read = Cpt(EpicsSignalRO, "Pck:B-I")
+    puck_dif_read = Cpt(EpicsSignalRO, "Pck:Dif-I")
+
+    # returns 0-16
+    # returns -1 if empty
+    samp_a_read = Cpt(EpicsSignalRO, "Samp:A-I")
+    samp_b_read = Cpt(EpicsSignalRO, "Samp:B-I")
+    samp_dif_read = Cpt(EpicsSignalRO, "Samp:Dif-I")
+
+    def dryGripper(self):
+        self.dry_traj.set(1)
+
+    def movement_ready(self):
+        if not self.power_sts.get():
+            return [False, "Power is off"]
+        if self.moving_sts.get():
+            return [False, "Moving"]
+        if self.paused_sts.get():
+            return [False, "Paused"]
+        return [True, "movement ready"]
+
+    # Recover Trajectory
+    #  required arguments
+    #  --current_tool_number
```

Comment:
The robot requires you to tell it what the currently equipped tool is when issuing trajectories. I could not see a reason other than forcing the person issuing the command to be absolutely certain it was the tool they want to be using.

Reply:
what I mean is that if that parameter, which looks like it is required for calling the function, is not actually required, then please re-write the documentation so that it makes sense. does it still make sense to you after a couple of weeks away from it?

Reply:
I have re-structured the comments here so that it is clear what part of the code requires these trajectory arguments and how they need to be provided. 6b200ff